### PR TITLE
SQL: TRUNCATE and ROUND functions

### LINKED ISSUE #33494: SQL: Implement ROUND and TRUNCATE math functions
Part of https://github.com/elastic/elasticsearch/issues/30007 for the remaining math functions.

### PR DESCRIPTION
**Pull request by astefan**
This adds the TRUNCATE math function and changes the ROUND one to accept two parameters. This fixes https://github.com/elastic/elasticsearch/issues/33494

### CODE REVIEW

File: x-pack/qa/sql/src/main/resources/math.csv-spec [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+---------------+---------------+-----------------+-----------------
+25976          |73717          |16               |49875
+29175          |73578          |20               |48164
+26436          |74999          |22               |52154
+27215          |74572          |18               |47733
+25324          |73851          |24               |44040
+;
+
+// https://github.com/elastic/elasticsearch/issues/33773
+//minMaxTruncateAndRoundOfAverageWithHavingRoundAndTruncate
```
Comment (costin) [inline x-pack/qa/sql/src/main/resources/math.csv-spec]: uncomment the test and add `-Ignored` to the its name to skip its execution

**Review by costin — CHANGES_REQUESTED** [review]
The docs need to be updated to include the new functions.
There's an issue for that already so it's up to you - I found it easier to add the doc with the PR since otherwise I forget about it.

Also the `getNumericFunctions` inside the JDBC driver need to be updated, to reflect the newly added functions.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/Truncate.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    protected ProcessorDefinition makeProcessorDefinition() {
+        return new BinaryMathProcessorDefinition(location(), this,
+                ProcessorDefinitions.toProcessorDefinition(left()),
+                ProcessorDefinitions.toProcessorDefinition(right()),
+                BinaryMathOperation.TRUNCATE);
+    }
+    
+    protected TypeResolution resolveInputType(DataType inputType) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/Truncate.java]: The method seems to be the exact copy of its parent - am I missing something?

Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/Truncate.java]: Not missing anything. Overlook on my side. Fixed.

**Comment by matriv** [comment]
I think it would be nice to keep also those functions with 1 argument. (second argument is implicitly 0 in that case). For postgres both trunc and round work this way: https://www.postgresql.org/docs/10/static/functions-math.html For MySQL/MariaDB only round works with 1 argument: https://mariadb.com/kb/en/library/round/ https://mariadb.com/kb/en/library/truncate/

What do you think?

**Comment by costin** [comment]
+1. We had `ROUND` with one parameter before, there's no reason not to have it.

**Comment by astefan** [comment]
@costin @matriv I've addressed your comments. Also, I've made the second parameter of these functions as optional.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java [hunk 3]
```diff
@@ -330,13 +332,17 @@ public boolean functionExists(String name) {
     static <T extends Function> FunctionDefinition def(Class<T> function,
             BinaryFunctionBuilder<T> ctorRef, String... aliases) {
         FunctionBuilder builder = (location, children, distinct, tz) -> {
-            if (children.size() != 2) {
+            boolean isBinaryOptionalParamFunction = function.isAssignableFrom(Round.class) || function.isAssignableFrom(Truncate.class);
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java]: The handling of 2 vs 1 parameter should not be handled here or at least this way - it should either be generic (so it doesn't care about the name) or is handled inside the Analyzer#ResolveFunctions. To not stall this any further, please raise another issue to have this as a follow-up.

File: x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt [hunk 4]
```diff
@@ -32,4 +32,6 @@ class org.elasticsearch.xpack.sql.expression.function.scalar.whitelist.InternalS
   String replace(String, String, String)
   Integer locate(String, String)
   Integer locate(String, String, Integer)
+  Number round(Number, Number)
```
Comment (costin) [inline x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt]: The declaration order should match that of the `FunctionRegistry` (the solved one - first based on category, then alphabetically).

**Review by costin — APPROVED** [review]
LGTM. Thanks for the docs. Let two comments but it's more for follow-up as oppose to the current issue.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java]: Acked. Issue created: https://github.com/elastic/elasticsearch/issues/33885

File: x-pack/qa/sql/src/main/resources/agg.sql-spec [hunk 5]
```diff
@@ -370,7 +370,7 @@ SELECT MIN(salary) mi, MAX(salary) ma, MAX(salary) - MIN(salary) AS d FROM test_
 aggHavingWithMultipleScalarFunctionsBasedOnAliasFromGroupBy
 SELECT MIN(salary) mi, MAX(salary) ma, MAX(salary) - MIN(salary) AS d FROM test_emp GROUP BY languages HAVING d - ma % mi > 0 AND AVG(salary) > 30000 ORDER BY languages;
 aggHavingWithMultipleScalarFunctionsBasedOnAliasFromGroupByAndAggNotInGroupBy
-SELECT MIN(salary) mi, MAX(salary) ma, MAX(salary) - MIN(salary) AS d FROM test_emp GROUP BY languages HAVING ROUND(d - ABS(ma % mi)) + AVG(salary) > 0 AND AVG(salary) > 30000 ORDER BY languages;
+SELECT MIN(salary) mi, MAX(salary) ma, MAX(salary) - MIN(salary) AS d FROM test_emp GROUP BY languages HAVING ROUND((d - ABS(ma % mi))) + AVG(salary) > 0 AND AVG(salary) > 30000 ORDER BY languages;
```
Comment (matriv) [inline x-pack/qa/sql/src/main/resources/agg.sql-spec]: Why do we need the extra parentheses here?

File: x-pack/qa/sql/src/main/resources/agg.sql-spec [hunk 6]
```diff
@@ -385,7 +385,7 @@ SELECT MIN(salary) mi, MAX(salary) ma, MAX(salary) - MIN(salary) AS d FROM test_
 aggMultiGroupByHavingWithMultipleScalarFunctionsBasedOnAliasFromGroupBy
 SELECT MIN(salary) mi, MAX(salary) ma, MAX(salary) - MIN(salary) AS d FROM test_emp GROUP BY gender, languages HAVING d - ma % mi > 0 AND AVG(salary) > 30000 ORDER BY gender, languages;
 aggMultiGroupByHavingWithMultipleScalarFunctionsBasedOnAliasFromGroupByAndAggNotInGroupBy
-SELECT MIN(salary) mi, MAX(salary) ma, MAX(salary) - MIN(salary) AS d FROM test_emp GROUP BY gender, languages HAVING ROUND(d - ABS(ma % mi)) + AVG(salary) > 0 AND AVG(salary) > 30000 ORDER BY gender, languages;
+SELECT MIN(salary) mi, MAX(salary) ma, MAX(salary) - MIN(salary) AS d FROM test_emp GROUP BY gender, languages HAVING ROUND((d - ABS(ma % mi))) + AVG(salary) > 0 AND AVG(salary) > 30000 ORDER BY gender, languages;
```
Comment (matriv) [inline x-pack/qa/sql/src/main/resources/agg.sql-spec]: and here?

**Review by matriv — APPROVED** [review]
LGTM. Nicely done. Left a question.

File: x-pack/qa/sql/src/main/resources/agg.sql-spec (hunk 5 above)
Comment (astefan) [reply x-pack/qa/sql/src/main/resources/agg.sql-spec]: @matriv overlooked aspect on my side. The previous version had the optional parameter `, 0)` and after making the second parameter optional, the brackets remained. I will remove them as part of an overall improvement to functions in https://github.com/elastic/elasticsearch/issues/33477.

Comment (matriv) [reply x-pack/qa/sql/src/main/resources/agg.sql-spec]: Ok! no worries. Thx!

**Comment by astefan** [comment]
6.x commit: https://github.com/elastic/elasticsearch/commit/78c26002de458b65cb9c8f279717aac21dc240af

**Pull request merged**